pkg/expr/benchmark_test.go

```diff
@@ -24,54 +24,61 @@ import (
 
 // run micro benchmark using expression evaluator.
 // Results vs hand coded expression eval function
-// 2017-03-10
+// 2017-03-16
 /*
-$ go test -bench=.
-BenchmarkExpression/ok_constantAST-8         	500000000	         3.70 ns/op
-BenchmarkExpression/ok_constantDirect-8      	1000000000	         2.43 ns/op
-BenchmarkExpression/ok_1stAST-8              	 1000000	      1626 ns/op
-BenchmarkExpression/ok_1stDirect-8           	100000000	        19.9 ns/op
-BenchmarkExpression/ok_2ndAST-8              	 1000000	      1623 ns/op
-BenchmarkExpression/ok_2ndDirect-8           	30000000	        41.5 ns/op
-BenchmarkExpression/ok_notfoundAST-8         	 1000000	      1628 ns/op
-BenchmarkExpression/ok_notfoundDirect-8      	30000000	        38.7 ns/op
+$ go test -run=^$  -bench=.  -benchmem
+LOR(EQ($a, 20), EQ(INDEX($request.header, "host"), "abc"))
+BenchmarkExpressionAST/ok_1stAST-8         	10000000	       188 ns/op	      16 B/op	       3 allocs/op
+BenchmarkExpressionAST/ok_2ndAST-8         	 3000000	       446 ns/op	      32 B/op	       5 allocs/op
+BenchmarkExpressionAST/ok_notfoundAST-8    	 3000000	       435 ns/op	      32 B/op	       5 allocs/op
+LOR(EQ($a, 20), EQ(INDEX($request.header, "host"), "abc"))
+BenchmarkExpressionDirect/ok_1stDirect-8   	100000000	        12.4 ns/op	       0 B/op	       0 allocs/op
+BenchmarkExpressionDirect/ok_2ndDirect-8   	50000000	        31.1 ns/op	       0 B/op	       0 allocs/op
+BenchmarkExpressionDirect/ok_notfoundDirect-8         	50000000	        29.5 ns/op	       0 B/op	       0 allocs/op
+PASS
+ok  	istio.io/mixer/pkg/expr	10.049s
 */
 
 type exprFunc func(a attribute.Bag) (bool, error)
 
 // a == 20 || request.header["host"] == 50
 
-func BenchmarkExpression(b *testing.B) {
-	success := "_SUCCESS_"
-	exprStr := `a == 20 || request.header["host"] == "abc"`
-	dff := func(a attribute.Bag) (bool, error) {
-		var aa int64
-		var b bool
-		var s map[string]string
-		if aa, b = a.Int64("a"); !b {
-			return false, fmt.Errorf("a not found")
-		}
-		if aa == 20 {
-			return true, nil
-		}
+func dff(a attribute.Bag) (bool, error) {
+	var aa int64
+	var b bool
+	var s map[string]string
+	if aa, b = a.Int64("a"); !b {
+		return false, fmt.Errorf("a not found")
+	}
+	if aa == 20 {
+		return true, nil
+	}
 
-		s, b = a.StringMap("request.header")
-		if !b {
-			return false, fmt.Errorf("a not found")
-		}
-		var ss string
-		ss = s["host"]
-		if ss == "abc" {
-			return true, nil
-		}
-		return false, nil
+	s, b = a.StringMap("request.header")
+	if !b {
+		return false, fmt.Errorf("a not found")
 	}
-	exf, _ := Parse(exprStr)
+	ss := s["host"]
+	if ss == "abc" {
```

Comment:
the choice of "abc" here seems a bit magical.  can we add some comments to help understanding what this function is doing and why these values are important?